pkg/pulumiyaml/config/config.go

```diff
@@ -120,3 +120,88 @@ func Parse(s string) (Type, bool) {
 		return nil, false
 	}
 }
+
+var (
+	ErrHeterogeneousList = HeterogeneousListErr{}
+	ErrEmptyList         = fmt.Errorf("empty list")
+	ErrUnexpectedType    = UnexpectedTypeErr{}
+)
+
+type HeterogeneousListErr struct {
+	T1 Type
+	T2 Type
+}
+
+func (e *HeterogeneousListErr) Error() string {
+	if e.T1 == nil || e.T2 == nil {
+		return "heterogeneous typed lists are not allowed"
+	}
+	return fmt.Sprintf("heterogeneous typed lists are not allowed: found types %s and %s",
+		e.T1, e.T2)
+}
+
+type UnexpectedTypeErr struct {
+	T interface{}
+}
+
+func (e *UnexpectedTypeErr) Error() string {
+	if e.T == nil {
+		return "unknown type"
+	}
+	return fmt.Sprintf("unexpected configuration type '%T': valid types are %s",
+		e.T, ConfigTypes,
+	)
+}
+
+// Type a go value into a configuration value.
+// If an error is returned, it is one of
+// - ErrHeterogeneousList
+// - ErrEmptyList
+// - ErrUnexpectedType
+func TypeValue(v interface{}) (Type, error) {
+	switch v := v.(type) {
+	case string:
+		return String, nil
+	case float64:
+		return Number, nil
+	case int:
+		return Int, nil
+	case bool:
+		return Boolean, nil
+	case []interface{}:
+		var expected Type
+		if len(v) == 0 {
+			return nil, ErrEmptyList
+		}
+		switch v[0].(type) {
+		case string:
+			expected = StringList
+		case float64:
+			expected = NumberList
+		case int:
+			expected = IntList
+		case bool:
+			expected = BooleanList
+		}
+		for i := 1; i < len(v); i++ {
+			if reflect.TypeOf(v[i-1]) != reflect.TypeOf(v[i]) {
+				t1, err := TypeValue(v[i-1])
+				if err != nil {
+					return nil, err
+				}
+				t2, err := TypeValue(v[i])
+				if err != nil {
+					return nil, err
+				}
+				return nil, &HeterogeneousListErr{t1, t2}
+			}
+		}
+		return expected, nil
+	case []float64:
```

Comment:
Do the cases for `[]float64` and `[]int` need to be above the case for `[]interface{}`?

Reply:
No. Go does not allow type coercion between `[]T` and `[]interface{}`.